File: tests/test_background.mjs
#!/usr/bin/env node
// Dependency-free regression test for the "make the background red" feature.
//
// Run: node tests/test_background.mjs
//
// Checks (against the repo root's index.html):
//   AC-2: the `body` rule's `background` declaration is red-dominant and no
//         longer contains the old purple gradient colors (#667eea / #764ba2).
//   AC-3: all other visual elements (mouse body/ears/eyes/nose/tail, trail
//         dots, heading/info text color) are byte-for-byte unchanged.
//   AC-4: the new red is dark enough that white text (h1 / .info) keeps a
//         WCAG contrast ratio of at least 4.5:1 against it.
//   AC-5: no other non-doc file in the repo still references the old
//         gradient colors.
//
// On any failure this prints every failing assertion with a clear message
// and exits 1. On success it prints a summary and exits 0.

import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const ROOT = path.resolve(__dirname, '..');
const INDEX_PATH = path.join(ROOT, 'index.html');

const failures = [];

function escapeRegex(s) {
  return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

/**
 * Find the body of a single CSS rule `selector { ... }` in a stylesheet
 * string, being careful not to match compound/descendant selectors that
 * merely contain `selector` as a substring (e.g. looking up `.eye` must not
 * match `.eye-left`, and `.trail-dot` must not match `.trail-dot.red`).
 */
function findRuleBlock(css, selector) {
  const esc = escapeRegex(selector);
  const isClass = selector.startsWith('.');
  const pre = isClass ? '(?<![\\w.-])' : '(?<![\\w-])';
  const post = '(?![\\w.-])';
  const re = new RegExp(pre + esc + post + '\\s*\\{([^}]*)\\}', 'i');
  const m = re.exec(css);
  return m ? m[1] : null;
}

function hexToRgb(hex) {
  let h = hex;
  if (h.length === 3) {
    h = h.split('').map((c) => c + c).join('');
  }
  const num = parseInt(h, 16);
  return { r: (num >> 16) & 255, g: (num >> 8) & 255, b: num & 255 };
}

// Best-effort support for CSS named colors in the red family, in case an
// implementation uses a named color instead of hex/rgb.
const NAMED_REDS = {
  red: [255, 0, 0],
  darkred: [139, 0, 0],
  firebrick: [178, 34, 34],
  crimson: [220, 20, 60],
  indianred: [205, 92, 92],
  maroon: [128, 0, 0],
  tomato: [255, 99, 71],
  orangered: [255, 69, 0],
  salmon: [250, 128, 114],
};

function matchNamedColor(value) {
  const lower = value.toLowerCase();
  for (const name of Object.keys(NAMED_REDS)) {
    const re = new RegExp(`(?<![\\w-])${name}(?![\\w-])`);
    if (re.test(lower)) {
      const [r, g, b] = NAMED_REDS[name];
      return { raw: name, r, g, b };
    }
  }
  return null;
}

/** Extract every color stop (hex, rgb()/rgba(), or a known red named color) from a CSS value. */
function parseColorStops(value) {
  const stops = [];

  const hexRe = /#([0-9a-fA-F]{3}|[0-9a-fA-F]{6})\b/g;
  let m;
  while ((m = hexRe.exec(value))) {
    stops.push({ raw: m[0], ...hexToRgb(m[1]) });
  }

  const rgbRe = /rgba?\(\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)/g;
  while ((m = rgbRe.exec(value))) {
    stops.push({ raw: m[0], r: Number(m[1]), g: Number(m[2]), b: Number(m[3]) });
  }

  if (stops.length === 0) {
    const named = matchNamedColor(value);
    if (named) stops.push(named);
  }

  return stops;
}

function srgbChannelToLinear(c) {
  const v = c / 255;
  return v <= 0.03928 ? v / 12.92 : Math.pow((v + 0.055) / 1.055, 2.4);
}

function relativeLuminance(r, g, b) {
  return (
    0.2126 * srgbChannelToLinear(r) +
    0.7152 * srgbChannelToLinear(g) +
    0.0722 * srgbChannelToLinear(b)
  );
}

/** WCAG contrast ratio of an rgb color against opaque white (#ffffff). */
function contrastWithWhite(r, g, b) {
  const L = relativeLuminance(r, g, b);
  const LWhite = 1.0;
  return (LWhite + 0.05) / (L + 0.05);
}

function walk(dir, relBase, excludedTop) {
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  let files = [];
  for (const entry of entries) {
    if (relBase === '' && excludedTop.has(entry.name)) continue;
    const full = path.join(dir, entry.name);
    const rel = relBase === '' ? entry.name : `${relBase}/${entry.name}`;
    if (entry.isDirectory()) {
      files = files.concat(walk(full, rel, excludedTop));
    } else if (entry.isFile()) {
      files.push({ full, rel });
    }
  }
  return files;
}

// ---------------------------------------------------------------------------

if (!fs.existsSync(INDEX_PATH)) {
  console.error(`FAIL: could not find index.html at ${INDEX_PATH}`);
  process.exit(1);
}

const html = fs.readFileSync(INDEX_PATH, 'utf8');

// --- AC-2: body background is red-dominant, and the old purple hexes are gone ---

let bodyStops = [];
const bodyBlock = findRuleBlock(html, 'body');
if (!bodyBlock) {
  failures.push('[AC-2] Could not find a `body { ... }` rule in index.html');
} else {
  const bgMatch = /(?<![\w-])background\s*:\s*([^;]+);/i.exec(bodyBlock);
  if (!bgMatch) {
    failures.push('[AC-2] Could not find a `background:` declaration inside the `body` rule');
  } else {
    const bgValue = bgMatch[1].trim();
    bodyStops = parseColorStops(bgValue);
    if (bodyStops.length === 0) {
      failures.push(
        `[AC-2] Could not parse any color stops (hex/rgb/known red name) from body background value: "${bgValue}"`
      );
    } else {
      const MARGIN = 15; // red channel must clearly exceed both other channels
      for (const s of bodyStops) {
        const dominant = s.r > s.g + MARGIN && s.r > s.b + MARGIN;
        if (!dominant) {
          failures.push(
            `[AC-2] body background color stop "${s.raw}" (rgb ${s.r}, ${s.g}, ${s.b}) is not clearly red-dominant ` +
              `(red channel must exceed both green and blue by a margin of at least ${MARGIN})`
          );
        }
      }
    }
  }
}

for (const oldColor of ['#667eea', '#764ba2']) {
  const re = new RegExp(escapeRegex(oldColor), 'i');
  if (re.test(html)) {
    failures.push(`[AC-2] index.html still contains the old gradient color ${oldColor}`);
  }
}

// --- AC-3: all other visual elements are byte-for-byte unchanged ---

const unchangedChecks = [
  { selector: '.mouse-body', prop: /background\s*:\s*#fff\s*;/i, desc: '.mouse-body { background: #fff; }' },
  { selector: '.ear-left', prop: /background\s*:\s*#fff\s*;/i, desc: '.ear-left { background: #fff; }' },
  { selector: '.ear-right', prop: /background\s*:\s*#fff\s*;/i, desc: '.ear-right { background: #fff; }' },
  { selector: '.tail', prop: /background\s*:\s*#fff\s*;/i, desc: '.tail { background: #fff; }' },
  { selector: '.eye', prop: /background\s*:\s*#333\s*;/i, desc: '.eye { background: #333; }' },
  { selector: '.nose', prop: /background\s*:\s*#ff6b9d\s*;/i, desc: '.nose { background: #ff6b9d; }' },
  {
    selector: '.trail-dot',
    prop: /background\s*:\s*rgba\(\s*59\s*,\s*130\s*,\s*246\s*,\s*0\.8\s*\)\s*;/i,
    desc: '.trail-dot { background: rgba(59, 130, 246, 0.8); }',
  },
  {
    selector: '.trail-dot.red',
    prop: /background\s*:\s*rgba\(\s*147\s*,\s*197\s*,\s*253\s*,\s*0\.8\s*\)\s*;/i,
    desc: '.trail-dot.red { background: rgba(147, 197, 253, 0.8); }',
  },
  { selector: 'h1', prop: /color\s*:\s*white\s*;/i, desc: 'h1 { color: white; }' },
  { selector: '.info', prop: /color\s*:\s*white\s*;/i, desc: '.info { color: white; }' },
];

for (const check of unchangedChecks) {
  const block = findRuleBlock(html, check.selector);
  if (!block) {
    failures.push(`[AC-3] Could not find a rule block for selector "${check.selector}"`);
    continue;
  }
  if (!check.prop.test(block)) {
    failures.push(
      `[AC-3] Expected "${check.desc}" to remain unchanged, but that declaration was not found in the "${check.selector}" rule`
    );
  }
}

// --- AC-4: white text stays legible (contrast >= 4.5:1) against the new red ---

if (bodyBlock && bodyStops.length === 0) {
  failures.push('[AC-4] Cannot verify text contrast because no color stops could be parsed from the body background');
} else {
  const MIN_CONTRAST = 4.5;
  for (const s of bodyStops) {
    const ratio = contrastWithWhite(s.r, s.g, s.b);
    if (ratio < MIN_CONTRAST) {
      failures.push(
        `[AC-4] body background color stop "${s.raw}" (rgb ${s.r}, ${s.g}, ${s.b}) has a contrast ratio of ` +
          `${ratio.toFixed(2)}:1 against white text, below the required ${MIN_CONTRAST}:1`
      );
    }
  }
}

// --- AC-5: no other non-doc file in the repo references the old gradient colors ---

const EXCLUDED_TOP_DIRS = new Set(['.git', 'docs', 'discovery', 'tests']);
const OLD_COLORS = ['#667eea', '#764ba2'];

let repoFiles = [];
try {
  repoFiles = walk(ROOT, '', EXCLUDED_TOP_DIRS);
} catch (err) {
  failures.push(`[AC-5] Failed to walk the repository looking for stale references: ${err.message}`);
}

for (const file of repoFiles) {
  let content;
  try {
    content = fs.readFileSync(file.full, 'utf8');
  } catch {
    continue; // unreadable (e.g. binary) file: skip
  }
  if (content.includes(' ')) continue; // binary heuristic: skip
  const lower = content.toLowerCase();
  for (const oldColor of OLD_COLORS) {
    if (lower.includes(oldColor.toLowerCase())) {
      failures.push(`[AC-5] File "${file.rel}" still references the old gradient color ${oldColor}`);
    }
  }
}

// ---------------------------------------------------------------------------

if (failures.length > 0) {
  console.error('FAIL: background regression checks failed:\n');
  for (const f of failures) console.error(`  - ${f}`);
  console.error(`\n${failures.length} assertion(s) failed.`);
  process.exit(1);
} else {
  console.log('PASS: all background regression checks passed (AC-2, AC-3, AC-4, AC-5).');
  process.exit(0);
}

File: tests/test_background.py
"""
Tests for the "make the background blue" feature.

These tests read the static `index.html` file directly (there is no build step
or test runner in this repo beyond the Python standard library) and assert on
the `body` rule's `background` declaration.

Run with:
    python3 -m unittest discover tests

NOTE: These tests are written against the *target* (post-change) state of
index.html, where the body background is a blue gradient. If run against the
pre-change baseline (still the purple/violet gradient
`linear-gradient(135deg, #667eea 0%, #764ba2 100%)`), assertions (1)-(3) below
are EXPECTED TO FAIL. That is intentional: it proves the test actually
exercises the acceptance criteria rather than trivially passing.
"""

import re
import unittest
from pathlib import Path


REPO_ROOT = Path(__file__).resolve().parent.parent
INDEX_HTML_PATH = REPO_ROOT / "index.html"

# Colors that must be gone from the body background once the fix lands.
FORBIDDEN_BACKGROUND_COLORS = ("#667eea", "#764ba2")

# Declarations that belong to *other* elements and must survive untouched.
UNTOUCHED_LITERALS = [
    "rgba(59, 130, 246, 0.8)",
    "rgba(147, 197, 253, 0.8)",
    "#ff6b9d",
    "#333",
]

UNTOUCHED_KEYFRAME_NAMES = ["pulse", "spin", "fadeOut", "blink", "swish"]

MIN_CONTRAST_RATIO = 4.5
WHITE_RGB = (255, 255, 255)


def _read_index_html():
    return INDEX_HTML_PATH.read_text(encoding="utf-8")


def _extract_body_rule(html):
    """
    Extract the contents of the CSS `body { ... }` rule (the top-level page
    background rule), being careful not to match rules like `.mouse-body {}`.

    We require `body` to appear as the entire selector on its line (only
    leading whitespace before it), so `.mouse-body {` is not mistaken for it.
    """
    match = re.search(r"^[ \t]*body[ \t]*\{([^}]*)\}", html, re.MULTILINE)
    if not match:
        raise AssertionError(
            "Could not find a top-level `body { ... }` CSS rule in index.html "
            "(searched for a line consisting of leading whitespace + 'body {')."
        )
    return match.group(1)


def _extract_background_declaration(body_rule_text):
    match = re.search(r"background\s*:\s*([^;]+);", body_rule_text)
    if not match:
        raise AssertionError(
            "The `body` rule has no `background: ...;` declaration to check."
        )
    return match.group(1).strip()


def _extract_color_stops(declaration):
    """
    Pull every color literal out of a CSS background declaration, returning
    a list of (label, (r, g, b)) tuples in the order they appear.

    Supports #rrggbb / #rgb hex colors and rgb()/rgba() functional notation,
    which covers every reasonable way to author a blue gradient/solid color
    for this rule.
    """
    stops = []

    hex_re = re.compile(r"#([0-9a-fA-F]{6}|[0-9a-fA-F]{3})\b")
    rgb_re = re.compile(
        r"rgba?\(\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)\s*(?:,\s*[\d.]+\s*)?\)"
    )

    for m in hex_re.finditer(declaration):
        hexval = m.group(1)
        if len(hexval) == 3:
            r, g, b = (int(c * 2, 16) for c in hexval)
        else:
            r = int(hexval[0:2], 16)
            g = int(hexval[2:4], 16)
            b = int(hexval[4:6], 16)
        stops.append((m.group(0), (r, g, b)))

    for m in rgb_re.finditer(declaration):
        r, g, b = int(m.group(1)), int(m.group(2)), int(m.group(3))
        stops.append((m.group(0), (r, g, b)))

    if not stops:
        raise AssertionError(
            "No parseable color stops (hex or rgb/rgba) were found in the "
            "body background declaration: %r" % (declaration,)
        )

    return stops


def _srgb_channel_to_linear(value_0_255):
    c = value_0_255 / 255.0
    if c <= 0.03928:
        return c / 12.92
    return ((c + 0.055) / 1.055) ** 2.4


def _relative_luminance(rgb):
    r, g, b = rgb
    r_lin = _srgb_channel_to_linear(r)
    g_lin = _srgb_channel_to_linear(g)
    b_lin = _srgb_channel_to_linear(b)
    return 0.2126 * r_lin + 0.7152 * g_lin + 0.0722 * b_lin


def _contrast_ratio(rgb_a, rgb_b):
    lum_a = _relative_luminance(rgb_a)
    lum_b = _relative_luminance(rgb_b)
    lighter = max(lum_a, lum_b)
    darker = min(lum_a, lum_b)
    return (lighter + 0.05) / (darker + 0.05)


class BackgroundIsBlueTests(unittest.TestCase):
    """AC-1: the body background must be blue, not the old purple/violet."""

    @classmethod
    def setUpClass(cls):
        cls.html = _read_index_html()
        cls.body_rule = _extract_body_rule(cls.html)
        cls.background_declaration = _extract_background_declaration(cls.body_rule)

    def test_old_purple_violet_colors_are_gone(self):
        for forbidden in FORBIDDEN_BACKGROUND_COLORS:
            self.assertNotIn(
                forbidden.lower(),
                self.background_declaration.lower(),
                msg=(
                    "The body background declaration still contains the old "
                    "purple/violet color %r (declaration: %r). AC-1 requires "
                    "this to be replaced with a blue background."
                    % (forbidden, self.background_declaration)
                ),
            )

    def test_every_background_color_stop_is_blue_dominant(self):
        stops = _extract_color_stops(self.background_declaration)
        for label, (r, g, b) in stops:
            self.assertTrue(
                b > r and b > g,
                msg=(
                    "Color stop %s (rgb=%r) in the body background is not "
                    "blue-dominant: blue channel (%d) must be strictly greater "
                    "than both red (%d) and green (%d) for the page to read as "
                    "blue rather than purple/violet/teal. Full declaration: %r"
                    % (label, (r, g, b), b, r, g, self.background_declaration)
                ),
            )

    def test_every_background_color_stop_has_sufficient_contrast_with_white(self):
        stops = _extract_color_stops(self.background_declaration)
        for label, rgb in stops:
            ratio = _contrast_ratio(rgb, WHITE_RGB)
            self.assertGreaterEqual(
                ratio,
                MIN_CONTRAST_RATIO,
                msg=(
                    "Color stop %s (rgb=%r) has a contrast ratio of only "
                    "%.2f:1 against white, which is below the WCAG AA "
                    "threshold of %.1f:1 required (AC-4) for readability of "
                    "the white mouse body and white h1/.info text against "
                    "this background. Full declaration: %r"
                    % (label, rgb, ratio, MIN_CONTRAST_RATIO, self.background_declaration)
                ),
            )


class UntouchedElementsTests(unittest.TestCase):
    """AC-2/AC-3: everything besides the body background must be unchanged."""

    @classmethod
    def setUpClass(cls):
        cls.html = _read_index_html()

    def test_unrelated_color_literals_are_still_present(self):
        for literal in UNTOUCHED_LITERALS:
            self.assertIn(
                literal,
                self.html,
                msg=(
                    "Expected the untouched declaration %r to still be present "
                    "verbatim in index.html, but it was not found. This "
                    "literal belongs to an element other than the body "
                    "background (mouse eyes, nose, or trail dots) and must "
                    "not be altered by a background-color-only change."
                    % (literal,)
                ),
            )

    def test_all_keyframe_animations_are_still_defined(self):
        for name in UNTOUCHED_KEYFRAME_NAMES:
            pattern = r"@keyframes\s+%s\b" % re.escape(name)
            self.assertRegex(
                self.html,
                pattern,
                msg=(
                    "Expected `@keyframes %s { ... }` to still be defined in "
                    "index.html so the corresponding animation keeps running, "
                    "but it was not found." % (name,)
                ),
            )


if __name__ == "__main__":
    unittest.main()
